# NFL Decision Tree Model
## Predicts the Superbowl using past data

In [1]:
import pandas as pd
import numpy as np
import os
working_directory = r"C:\Users\felix\Downloads\AI\SuperBowl Predicter"

In [2]:
os.chdir(working_directory)
print(os.getcwd())

C:\Users\felix\Downloads\AI\SuperBowl Predicter


### NFL Offenses

In [3]:
###NFL OFFensive Data from pro-football reference
#NFL2023OFF = pd.read_csv('NFL2023OFF.csv')
#NFL2023OFF['Year'] = '2023'

#NFL2022OFF = pd.read_csv('NFL2022OFF.csv')
#NFL2022OFF['Year'] = '2022'

#NFL2021OFF = pd.read_csv('NFL2021OFF.csv')
#NFL2021OFF['Year'] = '2021'

#NFL2020OFF = pd.read_csv('NFL2020OFF.csv')
#NFL2020OFF['Year'] = '2020'

NFL2019OFF = pd.read_csv('NFL2019OFF.csv')
NFL2019OFF['Year'] = '2019'

NFL2018OFF = pd.read_csv('NFL2018OFF.csv')
NFL2018OFF['Year'] = '2018'

NFL2017OFF = pd.read_csv('NFL2017OFF.csv')
NFL2017OFF['Year'] = '2017'

NFL2016OFF = pd.read_csv('NFL2016OFF.csv')
NFL2016OFF['Year'] = '2016'

NFL2015OFF = pd.read_csv('NFL2015OFF.csv')
NFL2015OFF['Year'] = '2015'

NFL2014OFF = pd.read_csv('NFL2014OFF.csv')
NFL2014OFF['Year'] = '2014'

NFL2013OFF = pd.read_csv('NFL2013OFF.csv')
NFL2013OFF['Year'] = '2013'

NFL2012OFF = pd.read_csv('NFL2012OFF.csv')
NFL2012OFF['Year'] = '2012'

NFL2011OFF = pd.read_csv('NFL2011OFF.csv')
NFL2011OFF['Year'] = '2011'

NFL2010OFF = pd.read_csv('NFL2010OFF.csv')
NFL2010OFF['Year'] = '2010'

NFL2009OFF = pd.read_csv('NFL2009OFF.csv')
NFL2009OFF['Year'] = '2009'

NFL2008OFF = pd.read_csv('NFL2008OFF.csv')
NFL2008OFF['Year'] = '2008'

NFL2007OFF = pd.read_csv('NFL2007OFF.csv')
NFL2007OFF['Year'] = '2007'

NFL2006OFF = pd.read_csv('NFL2006OFF.csv')
NFL2006OFF['Year'] = '2006'

NFL2005OFF = pd.read_csv('NFL2005OFF.csv')
NFL2005OFF['Year'] = '2005'

NFL2004OFF = pd.read_csv('NFL2004OFF.csv')
NFL2004OFF['Year'] = '2004'

NFL2003OFF = pd.read_csv('NFL2003OFF.csv')
NFL2003OFF['Year'] = '2003'

NFL2002OFF = pd.read_csv('NFL2002OFF.csv')
NFL2002OFF['Year'] = '2002'

NFL2001OFF = pd.read_csv('NFL2001OFF.csv')
NFL2001OFF['Year'] = '2001'

NFL2000OFF = pd.read_csv('NFL2000OFF.csv')
NFL2000OFF['Year'] = '2000'


In [4]:
# Merge response data together
data_frames = [NFL2018OFF[1:33], NFL2017OFF[1:33], NFL2016OFF[1:33], NFL2015OFF[1:33], NFL2014OFF[1:33], NFL2013OFF[1:33], NFL2012OFF[1:33],
              NFL2011OFF[1:33], NFL2010OFF[1:33], NFL2009OFF[1:33], NFL2008OFF[1:33], NFL2007OFF[1:33], NFL2006OFF[1:33], NFL2005OFF[1:33],
              NFL2004OFF[1:33], NFL2003OFF[1:33], NFL2002OFF[1:33], NFL2001OFF[1:32], NFL2000OFF[1:32]]
O_total = pd.concat(data_frames)

In [5]:
O_total.columns = ['O_Rank', 'Team', 'O_Games_Played', 'O_Points_For', 'O_Total_Yards',
       'O_Plays', 'O_Y/Play', 'O_TO', 'O_Fumbles_Lost',
       'O_1st_D', 'O_P_Completions', 'O_P_Attempts', 'O_P_Yards', 'O_P_TD',
       'O_P_Int', 'O_P_Y/PA', 'O_P_Passing_1st_D', 'O_R_Att', 'O_R_Yards',
       'O_R_TD', 'O_R_Y/A', 'O_R_1st_D', 'O_Pe', 'O_Pe_Yards',
       'O_Pe_1st_D', 'O_Scoring_Drives', 'O_TO_%', 'O_Expected_Points', 'Year']

O_total['Year'] = O_total['Year'].astype('int')

In [6]:
O_total

O_Rank                  Team O_Games_Played O_Points_For O_Total_Yards  \
1       1    Kansas City Chiefs             16          565          6810   
2       2      Los Angeles Rams             16          527          6738   
3       3    New Orleans Saints             16          504          6067   
4       4  New England Patriots             16          436          6295   
5       5    Indianapolis Colts             16          433          6179   
..    ...                   ...            ...          ...           ...   
27     27       Atlanta Falcons             16          252          3994   
28     28         Chicago Bears             16          216          4541   
29     29     Arizona Cardinals             16          210          4528   
30     30    Cincinnati Bengals             16          185          4260   
31     31      Cleveland Browns             16          161          3530   

   O_Plays O_Y/Play O_TO O_Fumbles_Lost O_1st_D  ... O_R_TD O_R_Y/A O_R_1st_D  \
1      996      6.8   18              6     384  ...     16     4.8       108   
2     1060      6.4   19              7     401  ...     23     4.9       134   
3     1010        6   16              9     377  ...     26     4.3       134   
4     1073      5.9   18              7     365  ...     18     4.3       131   
5     1070      5.8   24              9     371  ...     13     4.2        96   
..     ...      ...  ...            ...     ...  ...    ...     ...       ...   
27     925      4.3   34             14     256  ...      6     3.5        65   
28     993      4.6   29             13     239  ...      6     4.2        89   
29     932      4.9   44             20     253  ...      6     3.7        71   
30    1001      4.3   35             21     254  ...     13     4.7       119   
31     859      4.1   28              9     176  ...      7     3.2        53   

   O_Pe O_Pe_Yards O_Pe_1st_D O_Scoring_Drives O_TO_% O_Expected_Points  Year  
1   137       1152         37             52.6   10.5            298.14  2018  
2    96        878         31             48.6    9.8            239.79  2018  
3    94        939         33             51.8    9.5            237.73  2018  
4    93        744         28             40.8    9.2            190.96  2018  
5   120        953         38             41.2   13.2            140.47  2018  
..  ...        ...        ...              ...    ...               ...   ...  
27   97        720         35             24.6   16.9           -223.68  2000  
28   90        696          7             20.1   14.4           -250.25  2000  
29   96        785         26             21.6   22.2           -169.88  2000  
30   99        878         26             16.4   15.3           -232.53  2000  
31  103        925         13             15.8   13.7           -255.49  2000  

[606 rows x 29 columns]

In [7]:
#NFL2023DEF = pd.read_csv('NFL2023DEF.csv')
#NFL2023DEF['Year'] = '2023'

#NFL2022DEF = pd.read_csv('NFL2022DEF.csv')
#NFL2022DEF['Year'] = '2022'

#NFL2021DEF = pd.read_csv('NFL2021DEF.csv')
#NFL2021DEF['Year'] = '2021'

#NFL2020DEF = pd.read_csv('NFL2020DEF.csv')
#NFL2020DEF['Year'] = '2020'

NFL2019DEF = pd.read_csv('NFL2019DEF.csv')
NFL2019DEF['Year'] = '2019'

NFL2018DEF = pd.read_csv('NFL2018DEF.csv')
NFL2018DEF['Year'] = '2018'

NFL2017DEF = pd.read_csv('NFL2017DEF.csv')
NFL2017DEF['Year'] = '2017'

NFL2016DEF = pd.read_csv('NFL2016DEF.csv')
NFL2016DEF['Year'] = '2016'

NFL2015DEF = pd.read_csv('NFL2015DEF.csv')
NFL2015DEF['Year'] = '2015'

NFL2014DEF = pd.read_csv('NFL2014DEF.csv')
NFL2014DEF['Year'] = '2014'

NFL2013DEF = pd.read_csv('NFL2013DEF.csv')
NFL2013DEF['Year'] = '2013'

NFL2012DEF = pd.read_csv('NFL2012DEF.csv')
NFL2012DEF['Year'] = '2012'

NFL2011DEF = pd.read_csv('NFL2011DEF.csv')
NFL2011DEF['Year'] = '2011'

NFL2010DEF = pd.read_csv('NFL2010DEF.csv')
NFL2010DEF['Year'] = '2010'

NFL2009DEF = pd.read_csv('NFL2009DEF.csv')
NFL2009DEF['Year'] = '2009'

NFL2008DEF = pd.read_csv('NFL2008DEF.csv')
NFL2008DEF['Year'] = '2008'

NFL2007DEF = pd.read_csv('NFL2007DEF.csv')
NFL2007DEF['Year'] = '2007'

NFL2006DEF = pd.read_csv('NFL2006DEF.csv')
NFL2006DEF['Year'] = '2006'

NFL2005DEF = pd.read_csv('NFL2005DEF.csv')
NFL2005DEF['Year'] = '2005'

NFL2004DEF = pd.read_csv('NFL2004DEF.csv')
NFL2004DEF['Year'] = '2004'

NFL2003DEF = pd.read_csv('NFL2003DEF.csv')
NFL2003DEF['Year'] = '2003'

NFL2002DEF = pd.read_csv('NFL2002DEF.csv')
NFL2002DEF['Year'] = '2002'

NFL2001DEF = pd.read_csv('NFL2001DEF.csv')
NFL2001DEF['Year'] = '2001'

NFL2000DEF = pd.read_csv('NFL2000DEF.csv')
NFL2000DEF['Year'] = '2000'

In [8]:
# Merges response data together
data_frames_2 = [NFL2018DEF[1:33], NFL2017DEF[1:33], NFL2016DEF[1:33], NFL2015DEF[1:33], NFL2014DEF[1:33], NFL2013DEF[1:33], NFL2012DEF[1:33],
              NFL2011DEF[1:33], NFL2010DEF[1:33], NFL2009DEF[1:33], NFL2008DEF[1:33], NFL2007DEF[1:33], NFL2006DEF[1:33], NFL2005DEF[1:33],
              NFL2004DEF[1:33], NFL2003DEF[1:33], NFL2002DEF[1:33], NFL2001DEF[1:32], NFL2000DEF[1:32]]
D_total = pd.concat(data_frames_2)

In [9]:
D_total.columns = ['D_Rank', 'Team', 'D_Games_Played', 'D_Points_Allowed', 'D_Total_Yards',
       'D_Plays', 'D_Y/Play', 'D_TD', 'D_Fumbles_Lost',
       'D_1st_D', 'D_P_Completions', 'D_P_Attempts', 'D_P_Yards', 'D_P_TD',
       'D_P_Int', 'D_P_Y/PA', 'D_P_Passing_1st_D', 'D_R_Att', 'D_R_Yards',
       'D_R_TD', 'D_R_Y/A', 'D_R_1st_D', 'D_Pe', 'D_Pe_Yards',
       'D_Pe_1st_D', 'D_Scoring_Drives', 'D_TD_%', 'D_Expected_Points', 'Year']

D_total['Year'] = O_total['Year'].astype('int')

In [10]:
D_total.head()

D_Rank                  Team D_Games_Played D_Points_Allowed D_Total_Yards  \
1      1         Chicago Bears             16              283          4795   
2      2      Baltimore Ravens             16              287          4687   
3      3      Tennessee Titans             16              303          5334   
4      4        Houston Texans             16              316          5490   
5      5  Jacksonville Jaguars             16              316          4983   

  D_Plays D_Y/Play D_TD D_Fumbles_Lost D_1st_D  ... D_R_TD D_R_Y/A D_R_1st_D  \
1    1004      4.8   36              9     278  ...      5     3.8        67   
2     974      4.8   17              5     286  ...     11     3.7        82   
3    1001      5.3   17              6     298  ...      9     4.3        98   
4    1022      5.4   29             14     298  ...      8     3.4        76   
5     970      5.1   17              6     288  ...     16     4.3        89   

  D_Pe D_Pe_Yards D_Pe_1st_D D_Scoring_Drives D_TD_% D_Expected_Points  Year  
1  114       1022         25             28.6   19.5              95.5  2018  
2  113       1017         30             29.1    9.5             39.88  2018  
3  120        984         18             33.1    9.9            -22.13  2018  
4  117        945         22             31.6   13.9            -10.29  2018  
5   92        735         37             33.1    8.3              4.66  2018  

[5 rows x 29 columns]

In [11]:
trainSet = O_total.merge(D_total, on = ['Team', 'Year'])

#### Create Test Set

In [12]:
pd.set_option('display.max_columns', 500)

NFL2019OFF = pd.read_csv('NFL2019OFF.csv')
NFL2019OFF['Year'] = '2019'

NFL2019DEF = pd.read_csv('NFL2019DEF.csv')
NFL2019DEF['Year'] = '2019'

NFL2019OFF = NFL2019OFF[1:33]
NFL2019OFF.columns = ['O_Rank', 'Team', 'O_Games_Played', 'O_Points_For', 'O_Total_Yards',
       'O_Plays', 'O_Y/Play', 'O_TO', 'O_Fumbles_Lost',
       'O_1st_D', 'O_P_Completions', 'O_P_Attempts', 'O_P_Yards', 'O_P_TD',
       'O_P_Int', 'O_P_Y/PA', 'O_P_Passing_1st_D', 'O_R_Att', 'O_R_Yards',
       'O_R_TD', 'O_R_Y/A', 'O_R_1st_D', 'O_Pe', 'O_Pe_Yards',
       'O_Pe_1st_D', 'O_Scoring_Drives', 'O_TO_%', 'O_Expected_Points', 'Year']

NFL2019DEF = NFL2019DEF[1:33]
NFL2019DEF.columns = ['D_Rank', 'Team', 'D_Games_Played', 'D_Points_Allowed', 'D_Total_Yards',
       'D_Plays', 'D_Y/Play', 'D_TD', 'D_Fumbles_Lost',
       'D_1st_D', 'D_P_Completions', 'D_P_Attempts', 'D_P_Yards', 'D_P_TD',
       'D_P_Int', 'D_P_Y/PA', 'D_P_Passing_1st_D', 'D_R_Att', 'D_R_Yards',
       'D_R_TD', 'D_R_Y/A', 'D_R_1st_D', 'D_Pe', 'D_Pe_Yards',
       'D_Pe_1st_D', 'D_Scoring_Drives', 'D_TD_%', 'D_Expected_Points', 'Year']

testSet = NFL2019OFF.merge(NFL2019DEF, on = ['Team', 'Year'])
testSet.head()

O_Rank                  Team O_Games_Played O_Points_For O_Total_Yards  \
0      1      Baltimore Ravens             16          531          6521   
1      2   San Francisco 49ers             16          479          6097   
2      3  Tampa Bay Buccaneers             16          458          6366   
3      4    New Orleans Saints             16          458          5982   
4      5    Kansas City Chiefs             16          451          6067   

  O_Plays O_Y/Play O_TO O_Fumbles_Lost O_1st_D O_P_Completions O_P_Attempts  \
0    1064      6.1   15              7     386             289          440   
1    1012        6   23             10     336             331          478   
2    1086      5.9   41             11     353             382          630   
3    1011      5.9    8              2     347             418          581   
4     976      6.2   15             10     350             378          576   

  O_P_Yards O_P_TD O_P_Int O_P_Y/PA O_P_Passing_1st_D O_R_Att O_R_Yards  \
0      3225     37       8      6.9               171     596      3296   
1      3792     28      13      7.4               195     498      2305   
2      4845     33      30      7.2               244     409      1521   
3      4244     36       6        7               230     405      1738   
4      4498     30       5      7.5               211     375      1569   

  O_R_TD O_R_Y/A O_R_1st_D O_Pe O_Pe_Yards O_Pe_1st_D O_Scoring_Drives O_TO_%  \
0     21     5.5       188  109        867         27             52.1    8.6   
1     23     4.6       110  105        939         31             44.3     12   
2     15     3.7        81  133       1111         28             38.3   20.7   
3     12     4.3        97  120       1036         20               47    4.2   
4     16     4.2        93  107       1029         46             49.4      8   

  O_Expected_Points  Year D_Rank D_Games_Played D_Points_Allowed  \
0            245.99  2019      3             16              282   
1            123.17  2019      8             16              310   
2                44  2019     29             16              449   
3            167.04  2019     13             16              341   
4            217.32  2019      7             16              308   

  D_Total_Yards D_Plays D_Y/Play D_TD D_Fumbles_Lost D_1st_D D_P_Completions  \
0          4809     921      5.2   25             12     276             318   
1          4509     968      4.7   27             15     285             318   
2          5503    1073      5.1   28             16     331             408   
3          5329     999      5.3   23             10     311             371   
4          5594    1043      5.4   23              7     344             352   

  D_P_Attempts D_P_Yards D_P_TD D_P_Int D_P_Y/PA D_P_Passing_1st_D D_R_Att  \
0          544      3315     15      13      5.7               163     340   
1          519      2707     23      12      4.8               150     401   
2          664      4322     30      12      6.1               222     362   
3          603      3868     27      13      5.9               197     345   
4          582      3543     21      16      5.7               190     416   

  D_R_Yards D_R_TD D_R_Y/A D_R_1st_D D_Pe D_Pe_Yards D_Pe_1st_D  \
0      1494     12     4.4        74   97        795         39   
1      1802     11     4.5       105  116        957         30   
2      1181     11     3.3        70  111        971         39   
3      1461     12     4.2        71   92        713         43   
4      2051     14     4.9       115  116        844         39   

  D_Scoring_Drives D_TD_% D_Expected_Points  
0             32.9   14.6             16.61  
1               29   14.2            102.42  
2             39.6   13.5             12.23  
3             34.7   12.7            -41.89  
4             34.6   13.6            -54.58

In [13]:
adjusted_columns = ['O_Points_For', 'O_Total_Yards',
       'O_Plays', 'O_TO', 'O_Fumbles_Lost', 'O_1st_D',
       'O_P_Completions', 'O_P_Attempts', 'O_P_Yards', 'O_P_TD', 'O_P_Int',
       'O_P_Passing_1st_D', 'O_R_Att', 'O_R_Yards', 'O_R_TD',
       'O_R_1st_D', 'O_Pe', 'O_Pe_Yards', 'O_Pe_1st_D',
       'O_Scoring_Drives', 'O_Expected_Points', 
       'D_Games_Played', 'D_Points_Allowed', 'D_Total_Yards', 'D_Plays',
       'D_TD', 'D_Fumbles_Lost', 'D_1st_D', 'D_P_Completions',
       'D_P_Attempts', 'D_P_Yards', 'D_P_TD', 'D_P_Int',
       'D_P_Passing_1st_D', 'D_R_Att', 'D_R_Yards', 'D_R_TD', 
       'D_R_1st_D', 'D_Pe', 'D_Pe_Yards', 'D_Pe_1st_D', 'D_Scoring_Drives',
       'D_Expected_Points']

for col in [adjusted_columns]:
    testSet[col] = testSet[col].astype('float')

testSet['O_Games_Played'] = testSet['O_Games_Played'].astype('int')
adjuster = 16 / testSet['O_Games_Played'] 

testSet[adjusted_columns] = testSet[adjusted_columns].mul(adjuster, axis = 0)
testSet.head()

O_Rank                  Team  O_Games_Played  O_Points_For  O_Total_Yards  \
0      1      Baltimore Ravens              16         531.0         6521.0   
1      2   San Francisco 49ers              16         479.0         6097.0   
2      3  Tampa Bay Buccaneers              16         458.0         6366.0   
3      4    New Orleans Saints              16         458.0         5982.0   
4      5    Kansas City Chiefs              16         451.0         6067.0   

   O_Plays O_Y/Play  O_TO  O_Fumbles_Lost  O_1st_D  O_P_Completions  \
0   1064.0      6.1  15.0             7.0    386.0            289.0   
1   1012.0        6  23.0            10.0    336.0            331.0   
2   1086.0      5.9  41.0            11.0    353.0            382.0   
3   1011.0      5.9   8.0             2.0    347.0            418.0   
4    976.0      6.2  15.0            10.0    350.0            378.0   

   O_P_Attempts  O_P_Yards  O_P_TD  O_P_Int O_P_Y/PA  O_P_Passing_1st_D  \
0         440.0     3225.0    37.0      8.0      6.9              171.0   
1         478.0     3792.0    28.0     13.0      7.4              195.0   
2         630.0     4845.0    33.0     30.0      7.2              244.0   
3         581.0     4244.0    36.0      6.0        7              230.0   
4         576.0     4498.0    30.0      5.0      7.5              211.0   

   O_R_Att  O_R_Yards  O_R_TD O_R_Y/A  O_R_1st_D   O_Pe  O_Pe_Yards  \
0    596.0     3296.0    21.0     5.5      188.0  109.0       867.0   
1    498.0     2305.0    23.0     4.6      110.0  105.0       939.0   
2    409.0     1521.0    15.0     3.7       81.0  133.0      1111.0   
3    405.0     1738.0    12.0     4.3       97.0  120.0      1036.0   
4    375.0     1569.0    16.0     4.2       93.0  107.0      1029.0   

   O_Pe_1st_D  O_Scoring_Drives O_TO_%  O_Expected_Points  Year D_Rank  \
0        27.0              52.1    8.6             245.99  2019      3   
1        31.0              44.3     12             123.17  2019      8   
2        28.0              38.3   20.7              44.00  2019     29   
3        20.0              47.0    4.2             167.04  2019     13   
4        46.0              49.4      8             217.32  2019      7   

   D_Games_Played  D_Points_Allowed  D_Total_Yards  D_Plays D_Y/Play  D_TD  \
0            16.0             282.0         4809.0    921.0      5.2  25.0   
1            16.0             310.0         4509.0    968.0      4.7  27.0   
2            16.0             449.0         5503.0   1073.0      5.1  28.0   
3            16.0             341.0         5329.0    999.0      5.3  23.0   
4            16.0             308.0         5594.0   1043.0      5.4  23.0   

   D_Fumbles_Lost  D_1st_D  D_P_Completions  D_P_Attempts  D_P_Yards  D_P_TD  \
0            12.0    276.0            318.0         544.0     3315.0    15.0   
1            15.0    285.0            318.0         519.0     2707.0    23.0   
2            16.0    331.0            408.0         664.0     4322.0    30.0   
3            10.0    311.0            371.0         603.0     3868.0    27.0   
4             7.0    344.0            352.0         582.0     3543.0    21.0   

   D_P_Int D_P_Y/PA  D_P_Passing_1st_D  D_R_Att  D_R_Yards  D_R_TD D_R_Y/A  \
0     13.0      5.7              163.0    340.0     1494.0    12.0     4.4   
1     12.0      4.8              150.0    401.0     1802.0    11.0     4.5   
2     12.0      6.1              222.0    362.0     1181.0    11.0     3.3   
3     13.0      5.9              197.0    345.0     1461.0    12.0     4.2   
4     16.0      5.7              190.0    416.0     2051.0    14.0     4.9   

   D_R_1st_D   D_Pe  D_Pe_Yards  D_Pe_1st_D  D_Scoring_Drives D_TD_%  \
0       74.0   97.0       795.0        39.0              32.9   14.6   
1      105.0  116.0       957.0        30.0              29.0   14.2   
2       70.0  111.0       971.0        39.0              39.6   13.5   
3       71.0   92.0       713.0        43.0              34.7   12.7  

In [14]:
testSet.to_csv(r'C:\Users\felix\Downloads\AI\SuperBowl Predicter\TestSet3', index = False)


In [15]:
testSet['Year'] = testSet['Year'].astype('int')


mydata = pd.concat([testSet, trainSet])

In [16]:
mydata.head()

O_Rank                  Team O_Games_Played O_Points_For O_Total_Yards  \
0      1      Baltimore Ravens             16        531.0        6521.0   
1      2   San Francisco 49ers             16        479.0        6097.0   
2      3  Tampa Bay Buccaneers             16        458.0        6366.0   
3      4    New Orleans Saints             16        458.0        5982.0   
4      5    Kansas City Chiefs             16        451.0        6067.0   

  O_Plays O_Y/Play  O_TO O_Fumbles_Lost O_1st_D O_P_Completions O_P_Attempts  \
0  1064.0      6.1  15.0            7.0   386.0           289.0        440.0   
1  1012.0        6  23.0           10.0   336.0           331.0        478.0   
2  1086.0      5.9  41.0           11.0   353.0           382.0        630.0   
3  1011.0      5.9   8.0            2.0   347.0           418.0        581.0   
4   976.0      6.2  15.0           10.0   350.0           378.0        576.0   

  O_P_Yards O_P_TD O_P_Int O_P_Y/PA O_P_Passing_1st_D O_R_Att O_R_Yards  \
0    3225.0   37.0     8.0      6.9             171.0   596.0    3296.0   
1    3792.0   28.0    13.0      7.4             195.0   498.0    2305.0   
2    4845.0   33.0    30.0      7.2             244.0   409.0    1521.0   
3    4244.0   36.0     6.0        7             230.0   405.0    1738.0   
4    4498.0   30.0     5.0      7.5             211.0   375.0    1569.0   

  O_R_TD O_R_Y/A O_R_1st_D   O_Pe O_Pe_Yards O_Pe_1st_D O_Scoring_Drives  \
0   21.0     5.5     188.0  109.0      867.0       27.0             52.1   
1   23.0     4.6     110.0  105.0      939.0       31.0             44.3   
2   15.0     3.7      81.0  133.0     1111.0       28.0             38.3   
3   12.0     4.3      97.0  120.0     1036.0       20.0             47.0   
4   16.0     4.2      93.0  107.0     1029.0       46.0             49.4   

  O_TO_% O_Expected_Points  Year D_Rank D_Games_Played D_Points_Allowed  \
0    8.6            245.99  2019      3           16.0            282.0   
1     12            123.17  2019      8           16.0            310.0   
2   20.7              44.0  2019     29           16.0            449.0   
3    4.2            167.04  2019     13           16.0            341.0   
4      8            217.32  2019      7           16.0            308.0   

  D_Total_Yards D_Plays D_Y/Play  D_TD D_Fumbles_Lost D_1st_D D_P_Completions  \
0        4809.0   921.0      5.2  25.0           12.0   276.0           318.0   
1        4509.0   968.0      4.7  27.0           15.0   285.0           318.0   
2        5503.0  1073.0      5.1  28.0           16.0   331.0           408.0   
3        5329.0   999.0      5.3  23.0           10.0   311.0           371.0   
4        5594.0  1043.0      5.4  23.0            7.0   344.0           352.0   

  D_P_Attempts D_P_Yards D_P_TD D_P_Int D_P_Y/PA D_P_Passing_1st_D D_R_Att  \
0        544.0    3315.0   15.0    13.0      5.7             163.0   340.0   
1        519.0    2707.0   23.0    12.0      4.8             150.0   401.0   
2        664.0    4322.0   30.0    12.0      6.1             222.0   362.0   
3        603.0    3868.0   27.0    13.0      5.9             197.0   345.0   
4        582.0    3543.0   21.0    16.0      5.7             190.0   416.0   

  D_R_Yards D_R_TD D_R_Y/A D_R_1st_D   D_Pe D_Pe_Yards D_Pe_1st_D  \
0    1494.0   12.0     4.4      74.0   97.0      795.0       39.0   
1    1802.0   11.0     4.5     105.0  116.0      957.0       30.0   
2    1181.0   11.0     3.3      70.0  111.0      971.0       39.0   
3    1461.0   12.0     4.2      71.0   92.0      713.0       43.0   
4    2051.0   14.0     4.9     115.0  116.0      844.0       39.0   

  D_Scoring_Drives D_TD_% D_Expected_Points  
0             32.9   14.6             16.61  
1             29.0   14.2            102.42  
2             39.6   13.5             12.23  
3             34.7   12.7            -41.89  
4             34.6   13.6            -54.58

In [17]:
columns = ['O_Games_Played', 'O_Points_For', 'O_Total_Yards',
       'O_Plays', 'O_Y/Play', 'O_TO', 'O_Fumbles_Lost', 'O_1st_D',
       'O_P_Completions', 'O_P_Attempts', 'O_P_Yards', 'O_P_TD', 'O_P_Int',
       'O_P_Y/PA', 'O_P_Passing_1st_D', 'O_R_Att', 'O_R_Yards', 'O_R_TD',
       'O_R_Y/A', 'O_R_1st_D', 'O_Pe', 'O_Pe_Yards', 'O_Pe_1st_D',
       'O_Scoring_Drives', 'O_TO_%', 'O_Expected_Points', 
       'D_Games_Played', 'D_Points_Allowed', 'D_Total_Yards', 'D_Plays',
       'D_Y/Play', 'D_TD', 'D_Fumbles_Lost', 'D_1st_D', 'D_P_Completions',
       'D_P_Attempts', 'D_P_Yards', 'D_P_TD', 'D_P_Int', 'D_P_Y/PA',
       'D_P_Passing_1st_D', 'D_R_Att', 'D_R_Yards', 'D_R_TD', 'D_R_Y/A',
       'D_R_1st_D', 'D_Pe', 'D_Pe_Yards', 'D_Pe_1st_D', 'D_Scoring_Drives',
       'D_TD_%', 'D_Expected_Points']

for col in [columns]:
    mydata[col] = mydata[col].astype('float')

### Append Superbowl Winners

In [18]:
mydata['Superbowl'] = 0

In [19]:
SBwinners = pd.read_csv('SuperbowlWinners.csv')
SBwinners.head()

Date         SB                Winner  Pts                 Loser  \
0  Feb 3 2019  LIII (53)  New England Patriots   13      Los Angeles Rams   
1  Feb 4 2018   LII (52)   Philadelphia Eagles   41  New England Patriots   
2  Feb 5 2017    LI (51)  New England Patriots   34       Atlanta Falcons   
3  Feb 7 2016         50        Denver Broncos   24     Carolina Panthers   
4  Feb 1 2015  XLIX (49)  New England Patriots   28      Seattle Seahawks   

   Pts.1             MVP                        Stadium         City  \
0      3  Julian Edelman          Mercedez-Benz Stadium      Atlanta   
1     33      Nick Foles              U.S. Bank Stadium  Minneapolis   
2     28       Tom Brady                    NRG Stadium      Houston   
3     10      Von Miller                 Levi's Stadium  Santa Clara   
4     24       Tom Brady  University of Phoenix Stadium     Glendale   

        State  
0     Georgia  
1   Minnesota  
2       Texas  
3  California  
4     Arizona

In [20]:
SBwinners.Date = pd.to_datetime(SBwinners.Date)

In [21]:
SBwinners['Year'] = pd.DatetimeIndex(SBwinners['Date']).year
SBwinners = SBwinners[['Year', 'Winner']]
SBwinners.head()

Year                Winner
0  2019  New England Patriots
1  2018   Philadelphia Eagles
2  2017  New England Patriots
3  2016        Denver Broncos
4  2015  New England Patriots

In [22]:
SBwinners['Year'] = SBwinners['Year'] - 1

In [23]:
SBwinners.Year = SBwinners.Year.astype('object')

In [24]:
SBwinners['Superbowl'] = 1

In [25]:
SBwinners.Year = SBwinners.Year.astype('int')
mydata.Year = mydata.Year.astype('int')

test = pd.merge(mydata, SBwinners, left_on = ['Year','Team'], right_on = ['Year', 'Winner'], how = 'left')
test['Superbowl_y'].value_counts()


1.0    19
Name: Superbowl_y, dtype: int64

In [26]:
test.drop(['Superbowl_x', 'Winner'], axis=1, inplace = True)


test.Superbowl_y = test.Superbowl_y.fillna(0)
test.rename(columns={'Superbowl_y':'Superbowl'}, inplace=True)
test.head()

In [27]:
pd.set_option('display.max_rows', 500)
test.Superbowl_y = test.Superbowl_y.fillna(0)
test.Superbowl_y = test.Superbowl_y.astype('int')
test.rename(columns = {'Superbowl_y': 'Superbowl'}, inplace = True)

In [28]:
test.dtypes
test.Superbowl.sum()

19

# Modeling

In [29]:
import pylab as pl
import scipy.optimize as opt
from sklearn import preprocessing
%matplotlib inline 
import matplotlib.pyplot as plt

In [30]:
test

O_Rank                  Team  O_Games_Played  O_Points_For  O_Total_Yards  \
0        1      Baltimore Ravens            16.0         531.0         6521.0   
1        2   San Francisco 49ers            16.0         479.0         6097.0   
2        3  Tampa Bay Buccaneers            16.0         458.0         6366.0   
3        4    New Orleans Saints            16.0         458.0         5982.0   
4        5    Kansas City Chiefs            16.0         451.0         6067.0   
..     ...                   ...             ...           ...            ...   
631     27       Atlanta Falcons            16.0         252.0         3994.0   
632     28         Chicago Bears            16.0         216.0         4541.0   
633     29     Arizona Cardinals            16.0         210.0         4528.0   
634     30    Cincinnati Bengals            16.0         185.0         4260.0   
635     31      Cleveland Browns            16.0         161.0         3530.0   

     O_Plays  O_Y/Play  O_TO  O_Fumbles_Lost  O_1st_D  O_P_Completions  \
0     1064.0       6.1  15.0             7.0    386.0            289.0   
1     1012.0       6.0  23.0            10.0    336.0            331.0   
2     1086.0       5.9  41.0            11.0    353.0            382.0   
3     1011.0       5.9   8.0             2.0    347.0            418.0   
4      976.0       6.2  15.0            10.0    350.0            378.0   
..       ...       ...   ...             ...      ...              ...   
631    925.0       4.3  34.0            14.0    256.0            285.0   
632    993.0       4.6  29.0            13.0    239.0            304.0   
633    932.0       4.9  44.0            20.0    253.0            316.0   
634   1001.0       4.3  35.0            21.0    254.0            207.0   
635    859.0       4.1  28.0             9.0    176.0            278.0   

     O_P_Attempts  O_P_Yards  O_P_TD  O_P_Int  O_P_Y/PA  O_P_Passing_1st_D  \
0           440.0     3225.0    37.0      8.0       6.9              171.0   
1           478.0     3792.0    28.0     13.0       7.4              195.0   
2           630.0     4845.0    33.0     30.0       7.2              244.0   
3           581.0     4244.0    36.0      6.0       7.0              230.0   
4           576.0     4498.0    30.0      5.0       7.5              211.0   
..            ...        ...     ...      ...       ...                ...   
631         514.0     2780.0    14.0     20.0       4.8              156.0   
632         542.0     2805.0    12.0     16.0       4.9              143.0   
633         554.0     3250.0    16.0     24.0       5.5              156.0   
634         454.0     1946.0     6.0     14.0       3.8              109.0   
635         483.0     2445.0     9.0     19.0       4.7              110.0   

     O_R_Att  O_R_Yards  O_R_TD  O_R_Y/A  O_R_1st_D   O_Pe  O_Pe_Yards  \
0      596.0     3296.0    21.0      5.5      188.0  109.0       867.0   
1      498.0     2305.0    23.0      4.6      110.0  105.0       939.0   
2      409.0     1521.0    15.0      3.7       81.0  133.0      1111.0   
3      405.0     1738.0    12.0      4.3       97.0  120.0      1036.0   
4      375.0     1569.0    16.0      4.2       93.0  107.0      1029.0   
..       ...        ...     ...      ...        ...    ...         ...   
631    350.0     1214.0     6.0      3.5       65.0   97.0       720.0   
632    417.0     1736.0     6.0      4.2       89.0   90.0       696.0   
633    343.0     1278.0     6.0      3.7       71.0   96.0       785.0   
634    495.0     2314.0    13.0      4.7      119.0   99.0       878.0   
635    336.0     1085.0     7.0      3.2       53.0  103.0       925.0   

     O_Pe_1st_D  O_Scoring_Drives  O_TO_%  O_Expected_Points  Year D_Rank  \
0          27.0              52.1     8.6             245.99  2019      3   
1          31.0              44.3    12.0             123.17  2019      8   
2          28.0              38.3    20.7              44.00  2019     29   
3       

In [31]:
#split into values for test and train
test.O_Rank = test.O_Rank.astype('int')
test.D_Rank = test.D_Rank.astype('int')

train_columns = ['O_Rank', 'O_Points_For', 'O_Total_Yards',
       'O_Y/Play', 'O_TO', 'O_1st_D',
       'O_P_Completions', 'O_P_Yards', 'O_P_TD', 'O_P_Int',
       'O_P_Y/PA', 'O_P_Passing_1st_D', 'O_R_Yards', 'O_R_TD',
       'O_R_Y/A', 'O_R_1st_D', 'O_Pe', 'O_Pe_Yards', 
       'O_Scoring_Drives', 'D_Rank',
       'D_Points_Allowed', 'D_Total_Yards', 
       'D_Y/Play', 'D_TD', 'D_Fumbles_Lost', 'D_1st_D', 'D_P_Completions',
       'D_P_Yards', 'D_P_TD', 'D_P_Int', 'D_P_Y/PA',
       'D_P_Passing_1st_D', 'D_R_Yards', 'D_R_TD', 'D_R_Y/A',
       'D_R_1st_D', 'D_Pe', 'D_Pe_Yards', 'D_Scoring_Drives']


X = np.asarray(test[train_columns])
y = np.asarray(test['Superbowl'])
print(X.shape)
print(y.shape)

(636, 39)
(636,)


In [32]:
#normalize
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
#X[0:5]

In [33]:
teams = test.loc[ :31 , 'Team']

O_rank = test.loc[0:31, 'O_Rank']
D_rank = test.loc[0:31, 'D_Rank']

X_test = X[0:32]
X_train = X[32:]

y_test = y[0:32]
y_train = y[32:]

print(X_test.shape)
print(y_test.shape)
print(X_train.shape)
print(y_train.shape)

(32, 39)
(32,)
(604, 39)
(604,)


### Logistic Regression

In [34]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C=0.01, solver='liblinear').fit(X_train,y_train)
LR

LogisticRegression(C=0.01, solver='liblinear')

In [35]:
yhat = LR.predict(X_test)
yhat

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [36]:
from sklearn import metrics

yhat_binary = LR.predict(X_train)
print("Accuracy:",metrics.accuracy_score(y_train, yhat_binary))

Accuracy: 0.9685430463576159


In [37]:
import sys
import numpy
numpy.set_printoptions(threshold=sys.maxsize)

np.vstack((y_train, yhat_binary)).T

array([[0, 0],
       [0, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1,

In [38]:
df = pd.DataFrame({'Team':teams, 'Prediction':yhat}) 
df

Team  Prediction
0       Baltimore Ravens           0
1    San Francisco 49ers           0
2   Tampa Bay Buccaneers           0
3     New Orleans Saints           0
4     Kansas City Chiefs           0
5         Dallas Cowboys           0
6   New England Patriots           0
7      Minnesota Vikings           0
8       Seattle Seahawks           0
9       Tennessee Titans           0
10      Los Angeles Rams           0
11   Philadelphia Eagles           0
12       Atlanta Falcons           0
13        Houston Texans           0
14     Green Bay Packers           0
15     Arizona Cardinals           0
16    Indianapolis Colts           0
17         Detroit Lions           0
18       New York Giants           0
19     Carolina Panthers           0
20  Los Angeles Chargers           0
21      Cleveland Browns           0
22         Buffalo Bills           0
23       Oakland Raiders           0
24        Miami Dolphins           0
25  Jacksonville Jaguars           0
26   Pittsburgh Steelers           0
27        Denver Broncos           0
28         Chicago Bears           0
29    Cincinnati Bengals           0
30         New York Jets           0
31   Washington Redskins           0

In [39]:
yhat_prob = LR.predict_proba(X_test)
yhat_prob

array([[0.70463816, 0.29536184],
       [0.72923924, 0.27076076],
       [0.77189033, 0.22810967],
       [0.71937968, 0.28062032],
       [0.70232507, 0.29767493],
       [0.72813265, 0.27186735],
       [0.675172  , 0.324828  ],
       [0.73247968, 0.26752032],
       [0.76504245, 0.23495755],
       [0.76389253, 0.23610747],
       [0.75109317, 0.24890683],
       [0.76719112, 0.23280888],
       [0.76197103, 0.23802897],
       [0.7902093 , 0.2097907 ],
       [0.75051687, 0.24948313],
       [0.8134524 , 0.1865476 ],
       [0.77764766, 0.22235234],
       [0.80261128, 0.19738872],
       [0.81913156, 0.18086844],
       [0.82253572, 0.17746428],
       [0.76620423, 0.23379577],
       [0.79759163, 0.20240837],
       [0.73543599, 0.26456401],
       [0.78952633, 0.21047367],
       [0.82362413, 0.17637587],
       [0.78913875, 0.21086125],
       [0.75768194, 0.24231806],
       [0.77170987, 0.22829013],
       [0.7595607 , 0.2404393 ],
       [0.81412252, 0.18587748],
       [0.

In [40]:
df2 = pd.DataFrame({'Team':teams, 'Prediction':yhat_prob[:,1], 'O_Rank':O_rank, 'D_rank':D_rank }) 
df2

Team  Prediction  O_Rank  D_rank
0       Baltimore Ravens    0.295362       1       3
1    San Francisco 49ers    0.270761       2       8
2   Tampa Bay Buccaneers    0.228110       3      29
3     New Orleans Saints    0.280620       4      13
4     Kansas City Chiefs    0.297675       5       7
5         Dallas Cowboys    0.271867       6      11
6   New England Patriots    0.324828       7       1
7      Minnesota Vikings    0.267520       8       5
8       Seattle Seahawks    0.234958       9      22
9       Tennessee Titans    0.236107      10      12
10      Los Angeles Rams    0.248907      11      17
11   Philadelphia Eagles    0.232809      12      15
12       Atlanta Falcons    0.238029      13      23
13        Houston Texans    0.209791      14      19
14     Green Bay Packers    0.249483      15       9
15     Arizona Cardinals    0.186548      16      28
16    Indianapolis Colts    0.222352      17      18
17         Detroit Lions    0.197389      18      26
18       New York Giants    0.180868      19      30
19     Carolina Panthers    0.177464      20      31
20  Los Angeles Chargers    0.233796      21      14
21      Cleveland Browns    0.202408      22      20
22         Buffalo Bills    0.264564      23       2
23       Oakland Raiders    0.210474      24      24
24        Miami Dolphins    0.176376      25      32
25  Jacksonville Jaguars    0.210861      26      21
26   Pittsburgh Steelers    0.242318      27       6
27        Denver Broncos    0.228290      28      10
28         Chicago Bears    0.240439      29       4
29    Cincinnati Bengals    0.185877      30      25
30         New York Jets    0.219487      31      16
31   Washington Redskins    0.176648      32      27

In [41]:
df2 = df2.sort_values(by = ['Prediction'], ascending = False)
df2['Prediction_Rank'] = df2['Prediction'].rank(ascending = False)

df2

Team  Prediction  O_Rank  D_rank  Prediction_Rank
6   New England Patriots    0.324828       7       1              1.0
4     Kansas City Chiefs    0.297675       5       7              2.0
0       Baltimore Ravens    0.295362       1       3              3.0
3     New Orleans Saints    0.280620       4      13              4.0
5         Dallas Cowboys    0.271867       6      11              5.0
1    San Francisco 49ers    0.270761       2       8              6.0
7      Minnesota Vikings    0.267520       8       5              7.0
22         Buffalo Bills    0.264564      23       2              8.0
14     Green Bay Packers    0.249483      15       9              9.0
10      Los Angeles Rams    0.248907      11      17             10.0
26   Pittsburgh Steelers    0.242318      27       6             11.0
28         Chicago Bears    0.240439      29       4             12.0
12       Atlanta Falcons    0.238029      13      23             13.0
9       Tennessee Titans    0.236107      10      12             14.0
8       Seattle Seahawks    0.234958       9      22             15.0
20  Los Angeles Chargers    0.233796      21      14             16.0
11   Philadelphia Eagles    0.232809      12      15             17.0
27        Denver Broncos    0.228290      28      10             18.0
2   Tampa Bay Buccaneers    0.228110       3      29             19.0
16    Indianapolis Colts    0.222352      17      18             20.0
30         New York Jets    0.219487      31      16             21.0
25  Jacksonville Jaguars    0.210861      26      21             22.0
23       Oakland Raiders    0.210474      24      24             23.0
13        Houston Texans    0.209791      14      19             24.0
21      Cleveland Browns    0.202408      22      20             25.0
17         Detroit Lions    0.197389      18      26             26.0
15     Arizona Cardinals    0.186548      16      28             27.0
29    Cincinnati Bengals    0.185877      30      25             28.0
18       New York Giants    0.180868      19      30             29.0
19     Carolina Panthers    0.177464      20      31             30.0
31   Washington Redskins    0.176648      32      27             31.0
24        Miami Dolphins    0.176376      25      32             32.0

In [42]:
#df2.to_csv(r'C:\Users\gwats\Documents\UMBC Data Science\Jupyter\NFL Decision Tree\New_Predictions_log1.csv', index = False)


### ADABOOST

In [43]:
# Load libraries
from sklearn.ensemble import AdaBoostClassifier
from sklearn import datasets
# Import train_test_split function
from sklearn.model_selection import train_test_split
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics

In [44]:
# Load libraries
from sklearn.ensemble import AdaBoostClassifier

# Import Support Vector Classifier
from sklearn.svm import SVC
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
svc=SVC(probability=True, kernel='linear')

# Create adaboost classifer object
abc =AdaBoostClassifier(n_estimators=20, estimator=svc,learning_rate=1)

#abc = AdaBoostClassifier(n_estimators=30, learning_rate=1)

# Train Adaboost Classifer
model = abc.fit(X_train, y_train)

In [45]:
#Predict the response for test dataset
y_pred_binary = model.predict(X_train)
y_pred = model.predict_proba(X_test)
y_pred

array([[0.96247065, 0.03752935],
       [0.96486403, 0.03513597],
       [0.96666696, 0.03333304],
       [0.96643383, 0.03356617],
       [0.96449331, 0.03550669],
       [0.96483908, 0.03516092],
       [0.96142511, 0.03857489],
       [0.96605505, 0.03394495],
       [0.96902572, 0.03097428],
       [0.9680324 , 0.0319676 ],
       [0.96666053, 0.03333947],
       [0.96772249, 0.03227751],
       [0.96768181, 0.03231819],
       [0.9702656 , 0.0297344 ],
       [0.96822664, 0.03177336],
       [0.97251296, 0.02748704],
       [0.96830892, 0.03169108],
       [0.97106196, 0.02893804],
       [0.9715822 , 0.0284178 ],
       [0.97243395, 0.02756605],
       [0.96760197, 0.03239803],
       [0.97159622, 0.02840378],
       [0.96726517, 0.03273483],
       [0.97139115, 0.02860885],
       [0.97267944, 0.02732056],
       [0.97254964, 0.02745036],
       [0.9685541 , 0.0314459 ],
       [0.96972219, 0.03027781],
       [0.96894383, 0.03105617],
       [0.97166332, 0.02833668],
       [0.

In [46]:
df3 = pd.DataFrame({'Team':teams, 'Prediction':y_pred[:,1], 'O_Rank':O_rank, 'D_rank':D_rank }) 
df3['Prediction_Rank'] = df3['Prediction'].rank(ascending = False)
df3 = df3.sort_values(by = ['Prediction'], ascending = False)
df3

Team  Prediction  O_Rank  D_rank  Prediction_Rank
6   New England Patriots    0.038575       7       1              1.0
0       Baltimore Ravens    0.037529       1       3              2.0
4     Kansas City Chiefs    0.035507       5       7              3.0
5         Dallas Cowboys    0.035161       6      11              4.0
1    San Francisco 49ers    0.035136       2       8              5.0
7      Minnesota Vikings    0.033945       8       5              6.0
3     New Orleans Saints    0.033566       4      13              7.0
10      Los Angeles Rams    0.033339      11      17              8.0
2   Tampa Bay Buccaneers    0.033333       3      29              9.0
22         Buffalo Bills    0.032735      23       2             10.0
20  Los Angeles Chargers    0.032398      21      14             11.0
12       Atlanta Falcons    0.032318      13      23             12.0
11   Philadelphia Eagles    0.032278      12      15             13.0
9       Tennessee Titans    0.031968      10      12             14.0
14     Green Bay Packers    0.031773      15       9             15.0
16    Indianapolis Colts    0.031691      17      18             16.0
26   Pittsburgh Steelers    0.031446      27       6             17.0
28         Chicago Bears    0.031056      29       4             18.0
8       Seattle Seahawks    0.030974       9      22             19.0
27        Denver Broncos    0.030278      28      10             20.0
13        Houston Texans    0.029734      14      19             21.0
30         New York Jets    0.029416      31      16             22.0
17         Detroit Lions    0.028938      18      26             23.0
23       Oakland Raiders    0.028609      24      24             24.0
18       New York Giants    0.028418      19      30             25.0
21      Cleveland Browns    0.028404      22      20             26.0
29    Cincinnati Bengals    0.028337      30      25             27.0
19     Carolina Panthers    0.027566      20      31             28.0
15     Arizona Cardinals    0.027487      16      28             29.0
25  Jacksonville Jaguars    0.027450      26      21             30.0
24        Miami Dolphins    0.027321      25      32             31.0
31   Washington Redskins    0.025259      32      27             32.0

In [47]:
print("Accuracy:",metrics.accuracy_score(y_train, y_pred_binary))

Accuracy: 0.9685430463576159


### Notes

##### Decision Tree Base Classifier
n_estimators = 10, Accuracy: 0.9784768211920529
n_estimators = 20, Accuracy: 0.9966887417218543
n_estimators = 30, Accuracy: 1.0

#### SVC Base Classifier
n_estimators = 20, Accuracy: 0.9685430463576159 Patriots
n_estimators = 30, Accuracy: 0.9685430463576159 Raven
n_estimators = 50, Accuracy: 0.9685430463576159 Patriots



In [48]:
import sys
import numpy
numpy.set_printoptions(threshold=sys.maxsize)

np.vstack((y_train, y_pred_binary)).T

array([[0, 0],
       [0, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1,

In [49]:
df3.to_csv(r"C:\Users\felix\Downloads\AI\SuperBowl Predicter\New_Predictions1.csv", index = False)
